### PR TITLE
Local relative paths allow "./" prefix

### LINKED ISSUE #6190: Relative page URL generation doesn't support local folder relative paths prefixed with ./
APIs that accept relative paths for generating Page URLs, e.g. `<form asp-page="" />` and `RedirectToPage("")`, should support local folder relative URLs prefixed with `./`

Instead it appends it to the current page path and it ends up generating a URL that's app root relative with "page" in the querystring.

@pranavkm @rynowak

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #6190 by allowing `./` relative paths.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs [hunk 1]
```diff
@@ -457,6 +457,11 @@ private static object CalculatePageName(ActionContext actionContext, string page
                     throw new InvalidOperationException(Resources.FormatUrlHelper_RelativePagePathIsNotSupported(pageName));
                 }
 
+                if(pageName.StartsWith("./"))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: spacing

File: src/Microsoft.AspNetCore.Mvc.Core/Resources.resx [hunk 2]
```diff
@@ -404,6 +404,6 @@
     <value>No page named '{0}' matches the supplied values.</value>
   </data>
   <data name="UrlHelper_RelativePagePathIsNotSupported" xml:space="preserve">
-    <value>The relative page path '{0}' can only can only be used while executing a Razor Page. Specify a root relative path with a leading '/' to generate a URL outside of a Razor Page.</value>
+    <value>The relative page path '{0}' can only be used while executing a Razor Page. Specify a root relative path with a leading '/' to generate a URL outside of a Razor Page.</value>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Resources.resx]: Oops

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: `StringComparison.Ordinal`

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs [hunk 3]
```diff
@@ -457,6 +457,11 @@ private static object CalculatePageName(ActionContext actionContext, string page
...
+                {
+                    pageName = pageName.Substring(2, pageName.Length - 2);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: `pageName = pageName.Substring(2);`

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: Should we do back-slashes too? `ViewEnginePath` allows `..\` for path navigation.

File: test/WebSites/RazorPagesWebSite/Pages/Redirects/RedirectToSibling.cshtml [hunk 4]
```diff
@@ -8,4 +8,6 @@
     public IActionResult OnGetRedirectToSubDir() => RedirectToPage("SubDir/SubDirPage");
 
     public IActionResult OnGetRedirectToParent() => RedirectToPage("../Conventions/AuthFolder/Index");
+
+    public IActionResult OnGetRedirectToDotSlash() => RedirectToPage("./SubDir/SubDirPage");
```
Comment (ryanbrandenburg) [inline test/WebSites/RazorPagesWebSite/Pages/Redirects/RedirectToSibling.cshtml]: Double check my thinking that `/Pages/Redirects/SubDir/SubDirPage` is the correct outcome of this (as tested above). This seems correct to me but some instinct keeps wanting the correct outcome to be `/Pages/Redirects/RedirectToSibling/SubDir/SubDirPage`.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: test/WebSites/RazorPagesWebSite/Pages/Redirects/RedirectToSibling.cshtml (hunk 4 above)
Comment (rynowak) [inline test/WebSites/RazorPagesWebSite/Pages/Redirects/RedirectToSibling.cshtml]: Yep. The last segment always gets trimmed.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ViewEnginePath.cs [hunk 5]
```diff
@@ -70,6 +71,11 @@ public static string ResolvePath(string path)
                     }
                     pathSegments.RemoveAt(pathSegments.Count - 1);
                 }
+                else if (segment.Equals(CurrentDirectoryToken, StringComparison.Ordinal))
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ViewEnginePath.cs]: Can you add a test in https://github.com/aspnet/Mvc/blob/dev/test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorViewEngineTest.cs#L1475 if we're changing `ViewEnginePath`

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs [hunk 6]
```diff
@@ -923,8 +937,8 @@ public RazorPagesTest(MvcTestFixture<RazorPagesWebSite.Startup> fixture)
         public async Task TagHelpers_SupportSubDirectoryRoutes()
         {
             // Arrange
-var expected = 
-@"<form method=""post"" action=""/Pages/TagHelper/SubDir/SubDirPage""></form>
+            var expected =
+            @"<form method=""post"" action=""/Pages/TagHelper/SubDir/SubDirPage""></form>
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: Could you leave this untabbed?

Comment (ryanbrandenburg) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesTest.cs]: As discussed in person, leaving this because it's a VS formatting thing.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ViewEnginePath.cs (hunk 5 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ViewEnginePath.cs]: ViewEngine test?

Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ViewEnginePath.cs]: Added [here](https://github.com/aspnet/Mvc/pull/6202/files#diff-66cb8e5e8df1afdd1471d0372b32ff53R1475).

**Pull request merged**